src/Main.elm

```diff
@@ -1,10 +1,132 @@
 module Main exposing (main)
 
-import Data.Audience
-import Data.AudienceFolder
+import Data.Audience exposing ( Audience, subaudiences0 )
+import Data.AudienceFolder exposing ( AudienceFolder, subfolders0, folders0 )
 import Html exposing (Html)
+import Html.Events as Events
+import Html.Attributes as Attr
 
+import Browser
+
+import Dict exposing ( Dict )
+
+-- TODO: remove
+import Debug exposing ( log, todo, toString )
+
+-- ====== Overview ======
+-- 0. To compile the app, execute
+--      $ elm make src/Main.elm --output=public/main.js
+--    and open public/index.html. The css is in public/style.css and is not included when the app is executed via elm reactor.
+-- 1. Data.Audience and Data.AudienceFolder expose dictionaries
+--    whose keys are folder ids and whose values are folders, subfolders, and subaudience of the mentioned folder id.
+-- 2. If all parsing of JSON goes well, the state of the application is just
+--    a triple (record) of parent folder, its subfolders and subaudiences.
+-- 3. The only permissible action is to change the current parent folder.
+
+-- ===STATE=== --
+type alias ExplicitFolder =
+  { parentFolder : AudienceFolder
+  , subfolders : List AudienceFolder
+  , subaudiences : List Audience }
+
+type alias State = Maybe ExplicitFolder
+
+-- a helper function - not worth to put into a separate module
+listOfMaybeList : Maybe (List a) -> List a
+listOfMaybeList mxs =
+  case mxs of
+    Just xs -> xs
+    Nothing -> []
+
+stateOfFolderId : Int -> State
+stateOfFolderId folderId =
+  case (folders0, subfolders0, subaudiences0) of
+    (Ok foldersDict, Ok subfoldersDict, Ok subaudiencesDict) ->
+      Maybe.map
+        (\parentFolder -> 
+            ExplicitFolder
+              parentFolder
+              (listOfMaybeList (Dict.get folderId subfoldersDict))
+              (listOfMaybeList (Dict.get folderId subaudiencesDict))
+          )
+        (Dict.get folderId foldersDict)
+    _ -> Nothing
+
+initState : State
+initState = stateOfFolderId 357
+-- initState = stateOfFolderId 3111
+-- initState = Nothing
+
+-- ===ACTION===
+type Action = ChangeFolder Int
+
+action : Action -> State -> State
+action (ChangeFolder folderId) _ = stateOfFolderId folderId
```

Comment:
Unconventional naming! :)
Basically every Elm app I've seen does this as
```elm
type Msg = ...
update : Msg -> Model -> Model
```

Reply:
I think I chose `Action, State` naming because I was using React/Redux some time ago. Also we can think of `action : Action -> State -> State` as `\xs, s -> reduce action s xs : List[Action] -> State -> State`. In math this is usually called Monoid Action `(*) : M -> X -> X` of the Monoid `M` on a type `X`. But I am able to live with `update, Msg, Model` too :smile:

Reply:
Hey, I learned something here :) I didn't know the `M -> X -> X` thing had a name. `X -> X -> X` would ring a bell (magma, right? just a closed operation over X) but I didn't know about Monoid Actions. Thanks!